File: src/javascript/player.js
// interface between the network and the game
class Player {
	constructor(id) {
		this.brain = new Genome(genomeInputsN, genomeOutputsN, id);
		this.fitness;

		this.score = 1;
		this.lifespan = 0;
		this.dead = false;
		this.decisions = []; // current output values
		this.vision = []; // current input values
	}

	clone() {
		let clone = new Player();
		clone.brain = this.brain.clone();
		return clone;
	}

	crossover(parent) {
		let child = new Player();
		if (parent.fitness < this.fitness) {
			child.brain = this.brain.crossover(parent.brain);
		} else {
			child.brain = parent.brain.crossover(this.brain);
		}
		child.brain.mutate();
		return child;
	}

	// game stuff
	look() {
		// this.vision = [points[this.lifespan].x, points[this.lifespan].y];
		// this.correctVal = points[this.lifespan].type;

		//! input from the environment
		this.vision = [];
	}

	think() {
		this.decisions = this.brain.feedForward(this.vision);
	}

	move() {
		// let maxIndex = 0;
		// for (let i = 0; i < this.decisions.length; i++) {
		// 	if (this.decisions[i] > this.decisions[maxIndex]) {
		// 		maxIndex = i;
		// 	}
		// }
		// this.val = this.decisions[maxIndex] >= 0 ? 1 : 0;
		// TODO call the player movement
	}

	update() {
		// if (this.correctVal == this.val) {
		// 	this.score++;
		// }
		// this.lifespan++;
		// if (this.lifespan >= points.length) {
		// 	this.dead = true;
		// }
		//TODO update game stuff
		// update dead or alive, score, etc.
	}

	calculateFitness() {
		this.fitness = this.score;
		this.fitness /= this.brain.calculateWeight();
	}

	show() {
		// show the player
	}
}

File: src/cpp/player.cpp
#include "./header/player.h"
namespace neatCpp {
    // modify according to your needs
    void Player::calculateFitness() {
        fitness = score;
        fitness /= brain->calculateWeight();
    }
    // default implementation
    Player::Player(long int id, int inputSize, int outputSize) {
        brain = new Genome(inputSize, outputSize, id, false);
        fitness = 0;
        score = 1;
        lifespan = 0;
        dead = false;
        decisions.reserve(outputSize);
        vision.reserve(inputSize);
    }
    Player::Player(const Player* player) {
        brain = player->brain->clone();
        fitness = player->fitness;
        score = player->score;
        lifespan = player->lifespan;
        dead = player->dead;
        decisions = player->decisions;
        vision = player->vision;
    }
    Player::Player(Genome* _brain) {
        brain = _brain;
        fitness = 0;
        score = 1;
        lifespan = 0;
        dead = false;
        decisions.reserve(outputSize);
        vision.reserve(inputSize);
    }
    Player::~Player() {
        delete brain;
    }
    Player* Player::clone() const {
        return new Player(this);
    }
    Player* Player::crossover(const Player* const parent) {
        Genome* _brain;
        if (parent->fitness < fitness) {
            _brain = brain->crossover(parent->brain);
        } else {
            _brain = parent->brain->crossover(brain);
        }
        Player* child = new Player(_brain);
        child->brain->mutate();
        return child;
    }
    Genome* Player::getBrain() const { return brain; }
    long double Player::getFitness() const { return fitness; }
    void Player::setFitness(long double value) { fitness = value; }
    bool Player::isDead() const { return dead; }
    void Player::setDead(bool isDead) { dead = isDead; }
    long double Player::getScore() const { return score; }
    void Player::setScore(long double value) { score = value; }
    //
    void Player::look(const std::vector<long double>& inputVision) {
        vision = inputVision;
    }
    const std::vector<long double>& Player::think() {
        decisions = brain->feedForward(vision);
        return decisions;
    }
    const std::vector<long double>& Player::think(const std::vector<long double>& inputVision) {
        decisions = brain->feedForward(inputVision);
        return decisions;
    }

}